cpp-terminal/base.cpp

```diff
@@ -1,95 +1,427 @@
+        return 9;
+    } else if (rgb_compare(color, Bit4_reference::GREEN_BRIGHT) < 1) {
+        return 10;
+    } else if (rgb_compare(color, Bit4_reference::YELLOW_BRIGHT) < 1) {
+        return 11;
+    } else if (rgb_compare(color, Bit4_reference::BLUE_BRIGHT) < 1) {
+        return 12;
+    } else if (rgb_compare(color, Bit4_reference::MAGENTA_BRIGHT) < 1) {
+        return 13;
+    } else if (rgb_compare(color, Bit4_reference::CYAN_BRIGHT) < 1) {
+        return 14;
+    } else if (rgb_compare(color, Bit4_reference::WHITE_BRIGHT) < 1) {
+        return 15;
+    }
+
+    // check gray scale in 24 steps
+    if (color.r == color.g && color.r == color.b && color.g == color.b) {
+        return 232 + color.r / 32 + color.g / 32 + color.b / 32;
+    }
+
+    // normal color space
+    return 16 + 36 * (color.r / 51) + 6 * (color.g / 51) + (color.b / 51);
 }
 
-std::string Term::cursor_off() {
-    return "\x1b[?25l";
+bool Term::bit24_support() {
+    std::string colorterm = std::getenv("COLORTERM");
+    if (colorterm == "truecolor" || colorterm == "24bit") {
+        return true;
+    }
+    return false;
 }
 
-std::string Term::cursor_on() {
-    return "\x1b[?25h";
+std::string Term::rgb_to_bit24_auto_fg(RGB color) {
+    if (bit24_support()) {
+        return color_fg(color);
+    } else {
+        return color_fg(rgb_to_bit8(color));
+    }
 }
 
-std::string Term::clear_screen() {
-    return "\033[2J";
+std::string Term::rgb_to_bit24_auto_bg(RGB color) {
+    if (bit24_support()) {
+        return color_bg(color);
+    } else {
+        return color_bg(rgb_to_bit8(color));
+    }
 }
 
-std::string Term::clear_screen_buffer() {
-    return "\033[3J";
+/* FOREGROUND COLORS */
+
+std::string Term::color_fg(Term::Color4 color) {
+    return "\033[" + std::to_string((uint8_t)color + 30) + 'm';
 }
 
-std::string Term::move_cursor(size_t row, size_t col) {
-    return "\x1b[" + std::to_string(row) + ';' + std::to_string(col) + 'H';
+std::string Term::color_fg(uint8_t color) {
+    return "\033[38;5;" + std::to_string(color) + 'm';
+}
+std::string Term::color_fg(uint8_t r, uint8_t g, uint8_t b) {
+    return "\033[38;2;" + std::to_string(r) + ';' + std::to_string(g) + ';' +
+           std::to_string(b) + 'm';
 }
 
-std::string Term::move_cursor_right(int col) {
-    return "\x1b[" + std::to_string(col) + 'C';
+std::string Term::color_fg(Term::RGB rgb) {
+    if (rgb.empty) {
+        return color_fg(Color4::NONE);  // resets the current terminal color
+    }
+    return "\033[38;2;" + std::to_string(rgb.r) + ';' + std::to_string(rgb.g) +
+           ';' + std::to_string(rgb.b) + 'm';
+}
+std::string Term::color_fg(Term::RGBF rgbf) {
+    return color_fg(rgbf, rgbf.mode_fg);
 }
 
-std::string Term::move_cursor_down(int row) {
-    return "\x1b[" + std::to_string(row) + 'B';
+std::string Term::color_fg(Term::RGBF rgbf, Mode mode) {
+    switch (mode) {
+        case Mode::NONE:
```

Comment:
Maybe call it default?

Reply:
Term::Mode::NONE indicates that the given RGB color is empty, so it actually has none. It basically works like you make structs with FG/BG colors and then can on-runtime switch the struct values to change colors without needing to hard code them. Mode::NONE is important when you only set FG or BG colors or when you disable colors completely.

Reply:
ok :) UNSET could work too and maybe is closer to what you want to do?

Reply:
Yeah I guess so. Will change that ^^